sway/desktop/xwayland.c

```diff
@@ -666,6 +669,27 @@ static void handle_set_role(struct wl_listener *listener, void *data) {
 	view_execute_criteria(view);
 }
 
+static void handle_set_startup_id(struct wl_listener *listener, void *data) {
+	struct sway_xwayland_view *xwayland_view =
+		wl_container_of(listener, xwayland_view, set_startup_id);
+	struct sway_view *view = &xwayland_view->view;
+	struct wlr_xwayland_surface *xsurface = view->wlr_xwayland_surface;
+	if (xsurface->startup_id == NULL) {
+		return;
+	}
+
+	struct wlr_xdg_activation_token_v1 *token =
+		wlr_xdg_activation_v1_find_token(
+				server.xdg_activation_v1, xsurface->startup_id);
+	if (token == NULL) {
+		// We don't support external launchers in X
+		return;
+	}
+
+	struct launcher_ctx *ctx = token->data;
```

Comment:
This may be NULL: clients can create activation tokens as well.

Reply:
view_assign_ctx treats NULL as just a request to cleanup the currently existing context. I can see how that's not desirable though because it interferes with pid matching and prior activation requests, forcing the view to be placed on the active workspace when mapped.

I'll change this to return early. In a future PR I think I will add contexts to client tokens as well to support external launchers.